pkg/proxy/healthcheck/healthcheck.go

```diff
@@ -0,0 +1,291 @@
+		delete(h.serviceResponderMap, req.ServiceName)
+		(*sr.listener).Close()
+	}
+	// Create a service responder object and start listening and serving on the provided port
+	glog.V(2).Infof("Adding health check listener for service %s on nodePort %d", req.ServiceName, req.ListenPort)
+	server := http.Server{Addr: fmt.Sprintf(":%d", req.ListenPort), Handler: h}
+	ln, err := net.Listen("tcp", server.Addr)
+	if err != nil {
+		glog.Warningf("FAILED to listen on address %s (%s)", server.Addr, err)
+		return false
+	}
+	h.serviceResponderMap[req.ServiceName] = serviceResponder{serviceName: req.ServiceName,
+		listenPort: req.ListenPort,
+		listener:   &ln,
+		server:     &server}
+	go func(serviceName types.NamespacedName, p uint16, ln net.Listener, server http.Server) {
+		// Anonymous goroutine to block on Serve for this listen port - Serve will exit when the listener is closed
+		glog.V(2).Infof("Goroutine blocking on serving health checks for %s on port %d", serviceName, p)
+		if err := server.Serve(ln); err != nil {
+			glog.Infof("Proxy HealthCheck listen socket %d for service %s closed (%s)", p, serviceName, err)
+		}
+	}(req.ServiceName, req.ListenPort, ln, server)
+	return true
+}
+
+func (h *ProxyHC) handlerLoop() {
+	for {
+		// use separate channels for mutations vs health check servicing for fairness
+		select {
+		case req := <-h.hcRequestChannel:
+			h.handleHealthCheckRequest(req)
+		case req := <-h.mutationRequestChannel:
+			h.handleMutationRequest(req)
+		case req := <-h.listenerRequestChannel:
+			req.ResponseChannel <- h.handleServiceListenerRequest(req)
+		case <-h.shutdownChannel:
+			glog.Infof("Received kube-proxy LB health check handler loop shutdown request")
+			break
+		}
+	}
+}
+
+// Shutdown: Shutdown the main loop for this proxy health checker.
+func (h *ProxyHC) Shutdown() {
+	if h.shutdownChannel != nil {
+		h.shutdownChannel <- true
+	}
+}
+
+// UpdateEndpoints: Update the set of local endpoints for a service
+func (h *ProxyHC) UpdateEndpoints(serviceName types.NamespacedName, endpointUids sets.String) {
+	req := &ProxyMutationRequest{
+		ServiceName:  serviceName,
+		EndpointUids: &endpointUids,
+	}
+	h.mutationRequestChannel <- req
+}
+
+// UpdateServiceListener: Request addition/deletion of a service's health check listen port
+func (h *ProxyHC) UpdateServiceListener(serviceName types.NamespacedName, listenPort int, present bool) bool {
+	responseChannel := make(chan bool)
+	req := &ProxyListenerRequest{
+		ServiceName:     serviceName,
+		ListenPort:      uint16(listenPort),
+		Add:             present,
+		ResponseChannel: responseChannel,
+	}
+	h.listenerRequestChannel <- req
+	response := <-responseChannel
+	if !response {
+		glog.V(2).Infof("Failure updating service listener for service %s on port %d", serviceName, listenPort)
+	}
+	return response
+}
+
+// ProxyHealthCheckRequest - Factory method to instantiate the health check handler
+func ProxyHealthCheckFactory() *ProxyHC {
```

Comment:
nit: prefer just NewProxyHealthChecker or something to fit what we do elsewhere, factory feels like java